src/main/java/org/glite/authz/pep/pip/provider/policynamespip/UpdatingPolicyNamesCache.java

```diff
@@ -0,0 +1,210 @@
+    // instance variables
+    ////////////////////////////////////////////////////////////////////////
+ 
+    /** time interval (in msec) after which info files cache will be
+     * refreshed, default {@link #UPDATEINTERVAL}. */
+    private long update_interval = UPDATEINTERVAL;
+
+    /** Whether a thread is updating the {@link PolicyNamesCache} */
+    private volatile boolean updating = false;
+
+    /** Cache of info file directory
+     * @see PolicyNamesCache */
+    private volatile PolicyNamesCache cache = null;
+
+
+    ////////////////////////////////////////////////////////////////////////
+    // setter methods
+    ////////////////////////////////////////////////////////////////////////
+     
+    /**
+     * Sets the {@link #update_interval} (in msec) after which info files cache
+     * will be reprocessed.
+     * @param msecs number of millisecs between updates
+     * @see #UPDATEINTERVAL
+     */
+    public void setUpdateInterval(long msecs)    {
+	update_interval=msecs;
+    }
+    
+    ////////////////////////////////////////////////////////////////////////
+    // Constructors
+    ////////////////////////////////////////////////////////////////////////
+    
+    /**
+     * constructs new UpdatingPolicyNamesCache based on given trustDir and
+     * updateInterval
+     * @param trustDir directory containing info files
+     * @throws IOException on read errors for trust_dir or one of the info files
+     * @see #UpdatingPolicyNamesCache(String)
+     * @throws IOException in case of I/O errors
+     */
+    public UpdatingPolicyNamesCache(String trustDir, long updateInterval) throws IOException {
+	this(trustDir);
+	this.update_interval = updateInterval;
+    }
+
+    /**
+     * constructs new UpdatingPolicyNamesCache based on given trustDir and
+     * default {@link #UPDATEINTERVAL}.
+     * @param trustDir directory containing info files
+     * @param updateInterval interval (msec) between info file cache updates
+     * @see #UpdatingPolicyNamesCache(String, long)
+     * @throws IOException in case of I/O errors
+     */
+    public UpdatingPolicyNamesCache(String trustDir) throws IOException {
+	this.cache = new PolicyNamesCache(trustDir);
+    }
+
+
+    ////////////////////////////////////////////////////////////////////////
+    // Main instance method
+    ////////////////////////////////////////////////////////////////////////
+
+    /**
+     * Tries to find given subjectDN in the info files in trust dir.
+     * @param dn String subject DN to look for
+     * @return array of String with all the matching info files
+     * @throws IOException upon reading errors in updating the
+     * {@link PolicyNamesCache}
+     * @see PolicyNamesCache#matchIssuerDN(String)
+     */
+    public String[] findIssuerDN(String dn) throws IOException   {
+	// Update the cache (when needed)
+	updateCache();
+
+	// Obtain read lock
+	read.lock();
+	try {
+	    // Protect against empty cache
+	    if (cache == null)  {
```

Comment:
if you don't like NPEs, you should use an Optional for the cache.
I would drop this, since there's really no chance, if the code is written properly, that the cache is null.